WebDriverManager/Services/Impl/BinaryService.cs

```diff
@@ -7,45 +7,114 @@
+            zipPath = DownloadZip(url, zipPath);
+
+            //
+            // Copy or extract binary(s) into a staging directory
+            //
+            var stagingDir = Path.Combine(zipDir, "staging");
+            var stagingPath = Path.Combine(stagingDir, binaryName);
 
-            if (zipDestination.EndsWith(".exe", StringComparison.OrdinalIgnoreCase))
+            Directory.CreateDirectory(stagingDir);
+
+            if (zipPath.EndsWith(".exe", StringComparison.OrdinalIgnoreCase))
             {
-                File.Copy(zipDestination, binDestination);
+                File.Copy(zipPath, stagingPath);
             }
-            else if (zipDestination.EndsWith(".zip", StringComparison.OrdinalIgnoreCase))
+            else if (zipPath.EndsWith(".zip", StringComparison.OrdinalIgnoreCase))
             {
-                UnZip(zipDestination, binDestination, binaryName);
+                UnZip(zipPath, stagingPath, binaryName);
             }
-            else if (zipDestination.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
+            else if (zipPath.EndsWith(".tar.gz", StringComparison.OrdinalIgnoreCase))
             {
-                UnZipTgz(zipDestination, binDestination);
+                UnZipTgz(zipPath, stagingPath);
             }
 
 #if NETSTANDARD
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ||
-                RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            var needsChmod =
+                RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ||
+                RuntimeInformation.IsOSPlatform(OSPlatform.OSX);
+
+            if (needsChmod)
             {
-                var chmod = Process.Start("chmod", $"+x {binDestination}");
+                var chmod = Process.Start("chmod", $"+x {stagingPath}");
                 chmod?.WaitForExit();
             }
 #endif
 
-            RemoveZip(zipDestination);
-            return binDestination;
+            //
+            // Create the destination parent directory if it doesn't exist
+            //
+            var binaryParentDir = Path.GetDirectoryName(binaryDir);
+            Directory.CreateDirectory(binaryParentDir);
+
+            //
+            // Atomically rename the staging directory to the destination directory
+            //
+            // If a parallel thread or process races us and wins, the destination will already exist and this will fail
+            // with an IOException.
+            //
+            Exception renameException = null;
+            try
+            {
+                Directory.Move(stagingDir, binaryDir);
+            }
+            catch (Exception ex)
+            {
+                renameException = ex;
+            }
+
+            //
+            // Regardless of what happens, do a best-effort clean up
+            //
+            try
+            {
+                if (Directory.Exists(stagingDir)) Directory.Delete(stagingDir, true);
+            }
+            catch {}
```

Comment:
One more thing to do - can you please also process exceptions and not silently throw in such case.

Reply:
As the comment says, `Directory.Delete()`ing the temp directory is part of a "best-effort cleanup".  It doesn't matter whether it succeeds, so it shouldn't fail the entire program.

What about logging exception information to stderr if they occur?

Reply:
Hi @rosolko I've added code to write exception details to stderr if they occur.

Reply:
Thanks!